File: designui/src/core/lwwregister.ts
function nextTimestamp(): string {
    return Date.now().toString();
}

export class LWWRegister<T>{
    value:T
    timestamp:string
    counter:number

    constructor(initial:T, timestamp:string = nextTimestamp() + "-0"){
        this.value = initial;
        this.timestamp = timestamp;
        this.counter = 1;
    }

    get():T{
        return this.value;
    }

    set(value:T){
        this.timestamp = nextTimestamp() + "-" + this.counter.toString();
        this.value = value;
        this.counter += 1;
    }

    merge(other:LWWRegister<T>){
        if (this.timestamp <= other.timestamp){
            this.timestamp = other.timestamp;
            this.value = other.value;
        }
    }

    stateSnapshot():{value:T; timestamp:string}{
        return {
            value:this.value,
            timestamp:this.timestamp
        }
    }

    mergeStateSnapshot(other:{value:T; timestamp:string}){
        if(this.timestamp <= other.timestamp){
            this.timestamp = other.timestamp;
            this.value = other.value;
        }
    }
    
}

File: designserver/core/lwwregister.py
import time
def next_timestamp():
    return str(int(time.time() * 1000))

class LWWRegister:
    def __init__(self, initial, timestamp:str = next_timestamp() + "-0"):
        self.value = initial
        self.timestamp = timestamp
        self.counter = 1
    
    def get(self):
        return self.value
    
    def set(self, value):
        self.timestamp = next_timestamp() + "-" + str(self.counter)
        self.value = value
        self.counter += 1
    
    def merge(self, other:LWWRegister):
        if self.timestamp <= other.timestamp:
            self.timestamp = other.timestamp
            self.value = other.value
    
    def state_snapshot(self):
        return {"value":self.value, "timestamp":self.timestamp}
    
    def merge_state_snapshot(self, other):
        if self.timestamp <= other['timestamp']:
            self.timestamp = other['timestamp']
            self.value = other['value']